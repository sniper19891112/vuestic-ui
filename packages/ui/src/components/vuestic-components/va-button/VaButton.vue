<template>
  <component
    :is="tagComputed"
    class="va-button"
    :class="computedClass"
    :style="computedStyle"
    :disabled="disabled"
    :type="type"
    :href="hrefComputed"
    :target="target"
    :to="to"
    :replace="replace"
    :append="append"
    :active-class="activeClass"
    :exact="exact"
    :exact-active-class="exactActiveClass"
    @focus="updateFocusState(true)"
    @blur="updateFocusState(false)"
    :tabindex="loading ? -1 : 0"
  >
    <div
      class="va-button__content"
      v-on="inputListeners"
      @mouseenter="updateHoverState(true)"
      @mouseleave="updateHoverState(false)"
    >
      <va-progress-circle
        v-if="loading"
        indeterminate
        :size="loaderSize"
        :color="computedStyle.color"
        :thickness="0.15"
      />
      <template v-else>
        <va-icon
          v-if="icon"
          :name="icon"
          :size="$props.size"
          class="va-button__left-icon"
        />
        <slot />
        <va-icon
          v-if="iconRight"
          :name="iconRight"
          :size="$props.size"
          class="va-button__right-icon"
        />
      </template>
    </div>
  </component>
</template>

<script lang="ts">
import { inject, watchEffect, watch, Comment } from 'vue'
import { mixins, Vue, prop, Options, setup } from 'vue-class-component'
import {
  getGradientBackground,
  getFocusColor,
  getHoverColor,
  getBoxShadowColor,
} from '../../../services/color-config/color-functions'
import ColorMixin from '../../../services/color-config/ColorMixin'
import { RouterLinkMixin } from '../../vuestic-mixins/RouterLinkMixin/RouterLinkMixin'
import { SizeMixin } from '../../../mixins/SizeMixin'
import { LoadingMixin } from '../../vuestic-mixins/LoadingMixin/LoadingMixin'
import VaIcon from '../va-icon'
import { VaProgressCircle } from '../va-progress-bar'
import { ButtonGroupServiceKey } from '../va-button-group'

class ButtonProps {
  color = prop<string>({ type: String })
  textColor =prop<string>({ type: String, default: '#fff' })
  tag = prop<string>({ type: String, default: 'button' })
  outline = prop<boolean>({ type: Boolean, default: false })
  flat = prop<boolean>({ type: Boolean, default: false })
  type = prop<string>({ type: String, default: 'button' })
  disabled = prop<boolean>({ type: Boolean, default: false })
  block = prop<boolean>({ type: Boolean, default: false })
  round = prop<boolean>({ type: Boolean, default: true })
  equilateral = prop<boolean>({ type: Boolean, default: undefined })
  spaceBetweenItems = prop<boolean>({ type: Boolean, default: undefined })
  icon = prop<string>({ type: String, default: undefined })
  iconRight = prop<string>({ type: String, default: undefined })
  size = prop<string>({
    type: String,
    default: 'medium',
    validator: (value: string) => {
      return ['medium', 'small', 'large'].includes(value)
    },
  })
}

const ButtonPropsMixin = Vue.with(ButtonProps)

@Options({
  name: 'VaButton',
  components: { VaIcon, VaProgressCircle },
  emits: ['click'],
})
export default class VaButton extends mixins(
  ColorMixin,
  RouterLinkMixin,
  SizeMixin,
  LoadingMixin,
  ButtonPropsMixin,
) {
  hoverState = false
  focusState = false

  context = setup(() => {
    const buttonGroup = inject(ButtonGroupServiceKey, {})

    watch(() => this.$props.loading, (loading) => {
      if (loading) {
        this.updateFocusState(false)
        this.updateHoverState(false)
      }
    })

    watchEffect(() => {
      this.updateFocusState(this.hoverState)
      this.updateHoverState(this.hoverState)
    })

    return {
      buttonGroup,
    }
  })

  get colorComputed () {
    return this.computeColor(this.color, 'primary')
  }

  get hasDefaultSlot () {
    return this.$slots.default
  }

  get computedClass () {
    return {
      'va-button--default': !this.flat && !this.outline && !this.disabled,
      'va-button--flat': this.flat,
      'va-button--outline': this.outline,
      'va-button--disabled': this.disabled,
      'va-button--hover': this.hoverState,
      'va-button--focus': this.focusState,
      'va-button--with-left-content': this.$slots.prepend,
      'va-button--with-right-content': this.$slots.append,
      'va-button--large': this.size === 'large',
      'va-button--small': this.size === 'small',
      'va-button--normal': !this.size || this.size === 'medium',
      'va-button--loading': this.loading,
      'va-button--block': this.block,
      'va-button--square': !this.round,
      'va-button--equilateral': this.equilateral || !this.hasDefaultSlot || this.loading,
      'va-button--space-between-items': this.spaceBetweenItems,
    }
  }

  get buttonGroupColor () {
    return this.context?.buttonGroup?.color
  }

  get gradientStyle () {
    if (this.flat || this.outline) {
      return
    }
    // Allows button to grab color from button group.
    if (this.buttonGroupColor) {
      return
    }

    return getGradientBackground(this.colorComputed)
  }

  get shadowStyle () {
    if (this.flat || this.outline) {
      return
    }
    if (this.buttonGroupColor && this.theme.getColor(this.buttonGroupColor)) {
      return '0 0.125rem 0.19rem 0 ' +
        getBoxShadowColor(this.color ? this.colorComputed : this.theme.getColor(this.buttonGroupColor))
    }
    return '0 0.125rem 0.19rem 0 ' + getBoxShadowColor(this.colorComputed)
  }

  get loaderSize () {
    const size = /([0-9]*)(px)/.exec(this.sizeComputed) as null | [string, string, string]

    if (size) {
      return `${+size[1] / 2}${size[2]}`
    }

    return this.sizeComputed
  }

  get computedStyle () {
    const computedStyle: any = {
      color: '',
      borderColor: '',
      background: '',
      backgroundImage: '',
      boxShadow: '',
    }

    if (this.focusState) {
      if (this.outline || this.flat) {
        computedStyle.color = this.colorComputed
        computedStyle.borderColor = this.outline ? this.colorComputed : ''
        computedStyle.background = getFocusColor(this.colorComputed)
      } else {
        computedStyle.backgroundImage = this.gradientStyle
      }
    } else if (this.hoverState) {
      if (this.outline || this.flat) {
        computedStyle.color = this.colorComputed
        computedStyle.borderColor = this.outline ? this.colorComputed : ''
        computedStyle.background = getHoverColor(this.colorComputed)
      } else {
        computedStyle.backgroundImage = this.gradientStyle
        computedStyle.boxShadow = this.shadowStyle
      }
    } else {
      computedStyle.color = this.flat || this.outline ? this.colorComputed : this.textColor
      computedStyle.borderColor = this.outline ? this.colorComputed : ''
      computedStyle.backgroundImage = this.gradientStyle
      computedStyle.boxShadow = this.shadowStyle
    }

    if (this.buttonGroupColor && !this.outline && !this.flat) {
      computedStyle.background = this.$props.color ? this.colorComputed : this.theme.getColor(this.buttonGroupColor)
      computedStyle.backgroundImage = ''
    }

    return computedStyle
  }

  get inputListeners () {
    // vue3 $listeners.click -> $attrs.onClick
    return Object.assign({},
      this.$attrs,
      {
        click: (event: Event) => {
          this.$emit('click', event)
        },
      },
    )
  }

  updateHoverState (isHover: boolean) {
    this.hoverState = isHover
  }

  updateFocusState (isHover: boolean) {
    this.focusState = isHover
  }

  /** @public */
  focus (): void {
    (this.$el as HTMLElement).focus()
  }

  /** @public */
  blur (): void {
    console.log(this.$el);
    (this.$el as HTMLElement).blur()
  }
}
</script>

<style lang='scss'>
@import '../../vuestic-sass/resources/resources';
@import "variables";

.va-button {
  display: var(--va-button-display);
  align-items: var(--va-button-align-items);
  justify-content: var(--va-button-justify-content);
  background-image: var(--va-button-background-image);
  box-shadow: var(--va-button-box-shadow, var(--primary-control-box-shadow));
  outline: var(--va-button-outline);
  border: var(--va-button-border, var(--primary-control-border));
  font-family: var(--va-button-font-family, var(--primary-font-family));
  text-decoration: none !important;
  text-transform: initial;
  cursor: pointer;
  transition: var(--va-button-transition, var(--primary-transition));
  background-color: var(--va-button-background-color, var(--white));
  vertical-align: middle;
  box-sizing: border-box;

  &__content {
    display: flex;
    align-items: center;

    &__title,
    &__icon {
      display: flex;
      justify-content: center;
      align-items: center;
      margin: auto;
      white-space: nowrap;
    }
  }

  &--default {
    color: var(--va-button-background-color, var(--white));

    &:hover {
      opacity: 0.85;
    }

    &:focus,
    &:active {
      filter: brightness(85%);
    }

    i {
      color: var(--va-button-icon-color, var(--white));
    }
  }

  &--outline {
    background-color: transparent;
    border: solid var(--va-button-outline-border, var(--outline-border-width));
    text-decoration: none;

    &.va-button--disabled {
      background: transparent;

      @include va-disabled;

      &.va-button--active {
        .va-button__content,
        i {
          color: var(--va-button-outline-icon-color, var(--white)) !important;
        }
      }
    }
  }

  &--flat {
    background: transparent;
    border: var(--va-button-flat-border, var(--primary-control-border)) solid transparent;
    text-decoration: none;
  }

  &.va-button--disabled {
    @include va-disabled;
  }

  &--large {
    @include va-button(var(--va-button-lg-py), var(--va-button-lg-px), var(--va-button-lg-font-size), var(--va-button-lg-line-height), var(--va-button-lg-border-radius));

    letter-spacing: var(--va-button-lg-letter-spacing);
    height: 3rem;
    min-width: 3rem;

    .va-button__content {
      padding: var(--va-button-lg-py) var(--va-button-lg-px);
    }

    &.va-button--equilateral {
      width: 3rem;
    }

    &.va-button--outline {
      line-height: calc(var(--va-button-lg-line-height) - 2 * var(--va-button-outline-border, var(--outline-border-width)));
    }

    &.va-button--square {
      border-radius: var(--va-button-lg-square-border-radius);
    }
  }

  &--small {
    @include va-button(var(--va-button-sm-py), var(--va-button-sm-px), var(--va-button-sm-font-size), var(--va-button-sm-line-height), var(--va-button-sm-border-radius));

    letter-spacing: var(--va-button-sm-letter-spacing);
    height: 1.5rem;
    min-width: 1.5rem;

    .va-button__content {
      padding: var(--va-button-sm-py) var(--va-button-sm-px);
    }

    &.va-button--equilateral {
      width: 1.5rem;
    }

    &.va-button--outline {
      line-height: calc(var(--va-button-sm-line-height) - 2 * var(--va-button-outline-border, var(--outline-border-width)));
    }

    &.va-button--square {
      border-radius: var(--va-button-sm-square-border-radius);
    }
  }

  &--normal {
    @include va-button(var(--va-button-py), var(--va-button-px), var(--va-button-font-size), var(--va-button-line-height), var(--va-button-border-radius));

    letter-spacing: var(--va-button-letter-spacing, var(--primary-letter-spacing));
    height: 2.25rem;
    min-width: 2.25rem;

    .va-button__content {
      padding: var(--va-button-py) var(--va-button-px);
    }

    &.va-button--equilateral {
      width: 2.25rem;
    }

    &.va-button--outline {
      line-height: calc(var(--va-button-line-height) - 2 * var(--va-button-outline-border, var(--outline-border-width)));
    }

    &.va-button--square {
      border-radius: var(--va-button-square-border-radius);
    }
  }

  &--equilateral {
    .va-button__content { padding: 0; }
  }

  &--space-between-items {
    .va-button__content > * {
      margin-right: calc(var(--va-button-space-between-items) / 2);
      margin-left: calc(var(--va-button-space-between-items) / 2);
      &:last-child { margin-right: 0; }
      &:first-child { margin-left: 0; }
    }
  }

  &__left-icon {
    margin-right: calc(var(--va-button-space-between-items) / 2);
    &:last-child { margin-right: 0; }
  }

  &__right-icon {
    margin-left: calc(var(--va-button-space-between-items) / 2);
    &:last-child { margin-right: 0; }
  }

  &--loading {
    pointer-events: none;
  }

  &--block {
    display: flex;
    min-width: 100%;
  }

  &--square {
    border-radius: 0.5rem;
  }
}
</style>
