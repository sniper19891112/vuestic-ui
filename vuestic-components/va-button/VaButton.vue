<template>
  <component
    :is="computedTag"
    class="va-button"
    :class="buttonClass"
    :disabled="disabled"
    :type="type"
    :href="href"
    :target="target"
    :to="to"
    :replace="replace"
    :append="append"
    :active-class="activeClass"
    :exact="exact"
    :exact-active-class="exactActiveClass"
    v-on="inputListeners"
    tabindex="0"
  >
    <div class="va-button__content">
      <va-icon
        v-if="icon"
        fixed-width
        class="va-button__content__icon va-button__content__icon-left"
        :icon="icon"
      />
      <div
        v-if="hasTitleData"
        class="va-button__content__title">
        <slot/>
      </div>
      <va-icon
        v-if="iconRight"
        fixed-width
        class="va-button__content__icon va-button__content__icon-right"
        :icon="iconRight"
      />
    </div>
  </component>
</template>

<script>
export default {
  name: 'va-button',
  props: {
    tag: {
      type: String,
      default: 'button'
    },
    outline: {
      type: Boolean
    },
    flat: {
      type: Boolean
    },
    color: {
      type: String,
      default: 'success'
    },
    small: {
      type: Boolean
    },
    large: {
      type: Boolean
    },
    icon: {
      type: String
    },
    iconRight: {
      type: String
    },
    type: {
      type: String
    },
    disabled: {
      type: Boolean
    },
    /* Link props */
    href: {
      type: String
    },
    target: {
      type: String,
    },
    /* Router link props */
    to: {
      type: [String, Object]
    },
    replace: {
      type: Boolean
    },
    append: {
      type: Boolean
    },
    activeClass: {
      type: String
    },
    exact: {
      type: Boolean
    },
    exactActiveClass: {
      type: String
    }
  },
  computed: {
    buttonClass () {
      return {
        'va-button--success': this.color === 'success',
        'va-button--info': this.color === 'info',
        'va-button--danger': this.color === 'danger',
        'va-button--warning': this.color === 'warning',
        'va-button--gray': this.color === 'gray',
        'va-button--dark': this.color === 'dark',
        'va-button--default': !this.flat && !this.outline,
        'va-button--flat': this.flat,
        'va-button--outline': this.outline,
        'va-button--disabled': this.disabled,
        'va-button--without-title': !this.hasTitleData,
        'va-button--with-left-icon': this.icon,
        'va-button--with-right-icon': this.iconRight,
        'va-button--large': this.large,
        'va-button--small': this.small,
        'va-button--normal': !this.large && !this.small
      }
    },
    hasTitleData () {
      return this.$slots.default
    },
    computedTag () {
      if (this.tag === 'a' || this.href || this.target) {
        return 'a'
      }
      if (this.tag === 'router-link' || this.to || this.append || this.replace ||
        this.activeClass || this.exact || this.exactActiveClass) {
        return 'router-link'
      }
      return 'button'
    },
    inputListeners () {
      const vm = this
      return Object.assign({},
        this.$listeners,
        {
          click: function (event) {
            vm.$emit('click', event)
          }
        }
      )
    }
  },
}
</script>

<style lang='scss'>
  @mixin button-size($padding-y, $padding-x, $font-size, $line-height, $border-radius) {
    padding: $padding-y $padding-x;
    font-size: $font-size;
    line-height: $line-height;
    border-radius: $border-radius;
  }

  .va-button {
    display: inline-block;
    margin: $btn-margin;
    background-image: none;
    box-shadow: none;
    outline: none !important;
    border: $btn-border;
    font-family: $font-family-sans-serif;
    text-decoration: none !important;
    text-transform: initial;
    cursor: pointer;
    transition: $btn-transition;

    &__content {
      display: flex;

      &__title, &__icon {
        display: flex;
        justify-content: center;
        align-items: center;
        margin: auto;
      }
    }

    &--large {
      @include button-size($btn-padding-y-lg, $btn-padding-x-lg, $btn-font-size-lg, $btn-line-height-lg, $btn-border-radius-lg);
      letter-spacing: $btn-letter-spacing-lg;
      color: $white;

      .va-button__content__icon {
        width: $btn-icon-width-lg;
      }

      &.va-button--with-left-icon {
        padding-left: $btn-with-icon-wrapper-padding-lg;

        &.va-button--without-title {
          padding-right: $btn-with-icon-wrapper-padding-lg;
        }

        .va-button__content__title {
          padding-left: $btn-with-icon-content-padding-lg;
        }
      }

      &.va-button--with-right-icon {
        padding-right: $btn-with-icon-wrapper-padding-lg;

        .va-button__content__title {
          padding-right: $btn-with-icon-content-padding-lg;
        }
      }
    }

    &--small {
      @include button-size($btn-padding-y-sm, $btn-padding-x-sm, $btn-font-size-sm, $btn-line-height-sm, $btn-border-radius-sm);
      letter-spacing: $btn-letter-spacing-sm;
      color: $white;

      .va-button__content__icon {
        width: $btn-icon-width-sm;
      }

      &.va-button--with-left-icon {
        padding-left: $btn-with-icon-wrapper-padding-sm;

        &.va-button--without-title {
          padding-right: $btn-with-icon-wrapper-padding-sm;
        }

        .va-button__content__title {
          padding-left: $btn-with-icon-content-padding-sm;
        }
      }

      &.va-button--with-right-icon {
        padding-right: $btn-with-icon-wrapper-padding-sm;

        .va-button__content__title {
          padding-right: $btn-with-icon-content-padding-sm;
        }
      }
    }

    &--normal {
      @include button-size($btn-padding-y-nrm, $btn-padding-x-nrm, $btn-font-size-nrm, $btn-line-height-nrm, $btn-border-radius-nrm);
      letter-spacing: $btn-letter-spacing-nrm;
      color: $white;

      .va-button__content__icon {
        width: $btn-icon-width-nrm;
      }

      &.va-button--with-left-icon {
        padding-left: $btn-with-icon-wrapper-padding-nrm;

        &.va-button--without-title {
          padding-right: $btn-with-icon-wrapper-padding-nrm;
        }

        .va-button__content__title {
          padding-left: $btn-with-icon-content-padding-nrm;
        }
      }

      &.va-button--with-right-icon {
        padding-right: $btn-with-icon-wrapper-padding-nrm;

        .va-button__content__title {
          padding-right: $btn-with-icon-content-padding-nrm;
        }
      }
    }
  }

  $vuestic-colors: (
    success: (#63e5b3, #23e066, #77cea4, #40e583, #d6ffd3, #c0fbc7),
    danger: (#ff7455, #e34b4a, #b86e6d, #e34b4a, #ffebeb, #fbd2d2),
    warning: (#ffd72d, #feb900, #cbb06e, #ffc202, #fff3d1, #ffebb1),
    info: (#32b5e4, #2c82e0, #6c97ac, #2c82e0, #caeeff, #b2defb),
    gray: (#cdd0d5, #b4b6b9, #a3aab0, #babfc2, #e6e9ec, #dfe3e5),
    dark: (#576675, #34495e, #aebcca, #34495e, #afb6bb, #afb6bb)
  );

  @each $name, $colors in $vuestic-colors {
    $gradient-color1: nth($colors, 1);
    $gradient-color2: nth($colors, 2);
    $box-shadow: nth($colors, 3);
    $border-color: nth($colors, 4);
    $hover-color: nth($colors, 5);
    $focus-color: nth($colors, 6);

    .va-button--#{$name}.va-button--default{
      background-image: linear-gradient(to right, $gradient-color1, $gradient-color2);
      box-shadow: $btn-box-shadow $box-shadow;

      &:hover {
        background-image: linear-gradient(to right, lighten($gradient-color1, 15%), lighten($gradient-color2, 15%));
        color: $white !important;
      }

      &:active, &:focus, &.va-button--active {
        background-image: linear-gradient(to right, darken($gradient-color1, 15%), darken($gradient-color2, 15%));
        color: $white !important;
      }

      &.va-button--disabled {
        background-image: linear-gradient(to right, $gradient-color1, $gradient-color2);
        @include va-disabled;
      }
    }

    .va-button--#{$name}.va-button--outline{
      background: transparent;
      border: solid $btn-border-outline $border-color;
      text-decoration: none;
      color: $border-color !important;

      &:hover {
        background-color: $hover-color;
        color: $border-color !important;
      }

      &:active, &:focus, &.va-button--active {
        background-color: $focus-color;
        color: $border-color !important;

        .va-button-toggle & {
          background-color: $border-color;

          .va-button__content {
            color: $white;
          }

          i {
            color: $white;
          }
        }
      }

      &.va-button--disabled {
        background: transparent;
        @include va-disabled;
      }
    }

    .va-button--#{$name}.va-button--flat{
      background: transparent;
      border: $btn-border solid transparent;
      text-decoration: none;
      color: $border-color !important;

      &:hover {
        background-color: $hover-color;
        color: $border-color !important;
      }

      &:active, &:focus, &.va-button--active {
        background-color: $focus-color !important;
        color: $border-color !important;
        border: none;
      }

      &.va-button--disabled {
        background: transparent;
        @include va-disabled;
      }
    }

    .va-button--#{$name}.va-button--default{
      i {
        color: $white;
      }
    }

    .va-button--#{$name}.va-button--outline, .va-button--#{$name}.va-button--flat {
      i {
        color: $border-color;
      }
    }
  }
</style>
