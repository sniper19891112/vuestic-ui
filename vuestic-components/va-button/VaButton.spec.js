import { shallowMount, RouterLinkStub } from '@vue/test-utils'
import VaButton from './VaButton'

describe('VaButton', () => {
  /* 1. Button types */

  it('button without defined props', () => {
    const wrapper = shallowMount(VaButton)

    expect(wrapper.find(VaButton).classes()).toContain('va-button--default')
  })

  it('outline button', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: { outline: true }
    })
    expect(wrapper.find(VaButton).classes()).toContain('va-button--outline')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('flat button', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: { flat: true }
    })
    expect(wrapper.find(VaButton).classes()).toContain('va-button--flat')
    expect(wrapper.html()).toMatchSnapshot()
  })

  /* 2. Button sizes */

  it('normal button', () => {
    const wrapper = shallowMount(VaButton)

    expect(wrapper.find(VaButton).classes()).toContain('va-button--normal')
  })

  it('large button', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: { large: true }
    })
    expect(wrapper.find(VaButton).classes()).toContain('va-button--large')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('small button', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: { small: true }
    })
    expect(wrapper.find(VaButton).classes()).toContain('va-button--small')
    expect(wrapper.html()).toMatchSnapshot()
  })

  /* 3. Button colors */

  it('info button', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: { color: 'info' }
    })
    expect(wrapper.find(VaButton).classes()).toContain('va-button--info')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('danger button', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: { color: 'danger' }
    })

    expect(wrapper.find(VaButton).classes()).toContain('va-button--danger')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('warning button', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: { color: 'warning' }
    })

    expect(wrapper.find(VaButton).classes()).toContain('va-button--warning')
    expect(wrapper.html()).toMatchSnapshot()
  })

  /* 4. Button tags */

  it('default button', () => {
    const wrapper = shallowMount(VaButton)

    expect(wrapper.is('button')).toBe(true)
  })

  it('a button with defined href', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: {
        href: 'http://epic-spinners.epicmax.co/'
      }
    })
    expect(wrapper.is('a')).toBe(true)
  })

  it('a button with defined target', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: {
        target: '_blank'
      }
    })
    expect(wrapper.is('a')).toBe(true)
  })

  it('a button with defined href and target', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: {
        href: 'http://epic-spinners.epicmax.co/',
        target: '_blank'
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('router-link button with defined to property', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: {
        to: { name: 'charts' }
      },
      stubs: {
        RouterLink: RouterLinkStub
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('router-link button with defined to and active class properties', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: {
        to: { name: 'charts' },
        activeClass: 'va-button--active'
      },
      stubs: {
        RouterLink: RouterLinkStub
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  /* 5. Disabled buttons */

  it('disabled button', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: { disabled: true }
    })

    expect(wrapper.find(VaButton).classes()).toContain('va-button--disabled')
    expect(wrapper.html()).toMatchSnapshot()
  })

  /* 6. Buttons with icons */

  it('button with left icon', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: {
        icon: 'brandico brandico-facebook'
      }
    })
    expect(wrapper.find(VaButton).classes()).toContain('va-button--with-left-icon')
  })

  it('button with right icon', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: {
        iconRight: 'iconicstroke iconicstroke-info'
      }
    })
    expect(wrapper.find(VaButton).classes()).toContain('va-button--with-right-icon')
  })

  it('button with both icons', () => {
    const wrapper = shallowMount(VaButton, {
      propsData: {
        icon: 'brandico brandico-facebook',
        iconRight: 'iconicstroke iconicstroke-info'
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
