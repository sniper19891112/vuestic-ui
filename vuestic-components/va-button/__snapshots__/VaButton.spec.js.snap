// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaButton a button with defined href and target 1`] = `
<a href="http://epic-spinners.epicmax.co/" target="_blank" tabindex="0" class="va-button va-button--success va-button--default va-button--without-title va-button--normal">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</a>
`;

exports[`VaButton button with both icons 1`] = `
<button tabindex="0" class="va-button va-button--success va-button--default va-button--without-title va-button--with-left-icon va-button--with-right-icon va-button--normal">
  <div class="va-button__content">
    <va-icon fixed-width="" icon="brandico brandico-facebook" class="va-button__content__icon va-button__content__icon-left"></va-icon>
    <!---->
    <va-icon fixed-width="" icon="iconicstroke iconicstroke-info" class="va-button__content__icon va-button__content__icon-right"></va-icon>
  </div>
</button>
`;

exports[`VaButton danger button 1`] = `
<button tabindex="0" class="va-button va-button--danger va-button--default va-button--without-title va-button--normal">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</button>
`;

exports[`VaButton disabled button 1`] = `
<button disabled="disabled" tabindex="0" class="va-button va-button--success va-button--default va-button--disabled va-button--without-title va-button--normal">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</button>
`;

exports[`VaButton flat button 1`] = `
<button tabindex="0" class="va-button va-button--success va-button--flat va-button--without-title va-button--normal">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</button>
`;

exports[`VaButton info button 1`] = `
<button tabindex="0" class="va-button va-button--info va-button--default va-button--without-title va-button--normal">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</button>
`;

exports[`VaButton large button 1`] = `
<button tabindex="0" class="va-button va-button--success va-button--default va-button--without-title va-button--large">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</button>
`;

exports[`VaButton outline button 1`] = `
<button tabindex="0" class="va-button va-button--success va-button--outline va-button--without-title va-button--normal">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</button>
`;

exports[`VaButton router-link button with defined to and active class properties 1`] = `
<router-link to="[object Object]" active-class="va-button--active" tabindex="0" class="va-button va-button--success va-button--default va-button--without-title va-button--normal">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</router-link>
`;

exports[`VaButton small button 1`] = `
<button tabindex="0" class="va-button va-button--success va-button--default va-button--without-title va-button--small">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</button>
`;

exports[`VaButton warning button 1`] = `
<button tabindex="0" class="va-button va-button--warning va-button--default va-button--without-title va-button--normal">
  <div class="va-button__content">
    <!---->
    <!---->
    <!---->
  </div>
</button>
`;
